Implement Selective Export of Specific Tables from Database :  #13 export few tables

Josel099 opened the pull request:
### Overview
This pull request introduces a new feature that allows users to export specific tables from the database by specifying their names as a comma-separated list. 

### Changes Made
- Added a property key for specifying specific tables to export.
- Implemented logic to generate SQL statements for the specified tables.
- Updated documentation to include instructions on how to use the new feature.

### Example Usage
Users can now set the property for specific tables as follows:
```java
properties.setProperty(MysqlExportService.SPECIFIC_TABLES_FOR_EXPORT, "table_1,table_2,table_3");
```

@SeunMatt  **_Could you check if this approach is okay with you? If so, I'd also like to contribute to the section on importing specific tables from the dump file._**

SeunMatt committed:
Merge pull request #15 from SeunMatt/development

Development v1.2.1

SeunMatt committed:
Merge pull request #39 from SeunMatt/development

development to master

SeunMatt committed:
Merge pull request #42 from SeunMatt/development

Development

SeunMatt committed:
Merge pull request #44 from SeunMatt/development

Development

Commit:
feat: Implement specific tables export functionality

Added support for exporting only specified tables from the database. The implementation checks for a list of specific tables to be exported and generates the corresponding SQL statements for table creation and data insertion.

Commit:
docs: Update README.md

Commit:
refactor : format the code

Josel099 committed:
Merge branch 'development' into feature/Issue-13-export-few-tables

SeunMatt commented on src/main/java/com/smattme/MysqlExportService.java:
```diff
@@ -338,14 +347,32 @@ private String exportToSql() throws SQLException {
-                logger.error("Exception occurred while processing table: " + s, e);
+
+        // Retrieve specific tables from properties, if they exist
+        String specificTables = properties.containsKey(SPECIFIC_TABLES_FOR_EXPORT)
```
We can simplify this statement by using `String specificTables = properties.getProperty(SPECIFIC_TABLES_FOR_EXPORT, "");` instead of the existing ternary operator

Josel099 replied:
Good suggestion!

SeunMatt commented on src/main/java/com/smattme/MysqlExportService.java:
```diff
@@ -338,14 +347,32 @@ private String exportToSql() throws SQLException {
+            : "";
+
+        // Split the specific tables into a list
+        List<String> specificTableList = new ArrayList<>();
+        if (!specificTables.isEmpty()) {
+            specificTableList = Arrays.stream(specificTables.split(",")).map(String::trim).toList();
+        }
```
```suggestion
        List<String> specificTableList = new ArrayList<>();
        if (!specificTables.isEmpty()) {
            specificTableList = Arrays.stream(specificTables.split(",")).map(String::trim).toList();
        }
```

We can replace the above snippet with a ternary operator. As in:

```
List<String> specificTableList = !specificTables.isEmpty()
                                 ? Arrays.stream(specificTables.split(",")).map(String::trim).collect(Collectors.toList())
                                 : Collections.emptyList();
```

SeunMatt commented on src/main/java/com/smattme/MysqlExportService.java:
```diff
@@ -338,14 +347,32 @@ private String exportToSql() throws SQLException {
+
+            // Determine if the specific table list is empty or includes the current table name
+            // If so, generate SQL statements for table creation and data insertion
+            if (specificTableList.isEmpty() || specificTableList.contains(specificTableName)) {
```
This logic is better placed in the function `com.smattme.MysqlBaseService#getAllTablesAndViews`. Since the function is the one responsible for determining which tables will be added to list of `tables` for processing or not.

```
ResultSet rs;
        rs = stmt.executeQuery("SHOW TABLE STATUS FROM `" + database + "`;");
        while ( rs.next() ) {
           String name = rs.getString("Name");
           //TODO check if the table/view name is in the specified list OR the list is empty before adding them to the final list
            String comment = rs.getString("Comment");
            if("VIEW".equals(comment)) {
                views.add(name);
            }
            else {
                tables.add(name);
            }
        }
```

This means you'll need to modify the signature of the com.smattme.MysqlBaseService#getAllTablesAndViews to expect a `Properties` object. Then, in it, you'll construct a table whitelist and use it in the while loop for building the table.

**It would be even more awesome if it is possible to add the whitelist to the `SHOW TABLE STATUS FROM database` query directly**.

Looking forward to your update

SeunMatt reviewed (commented):
Kindly rebase your clone with the master branch so you can have the automated test code. Kindly add a test case for your new feature implementation as well.

Thank you for your contribution. I have made some suggestions and I am looking forward to the revised PR.

Josel099 committed:
Merge branch 'SeunMatt:master' into feature/Issue-13-export-few-tables

Josel099 committed:
Refactor : refactor getAllTablesAndViews to filter tables based on provided list

- Moved filtering logic from exportToSql into getAllTablesAndViews.
- Now, if a non-empty specificTableList is provided, only tables/views in that list are returned.
- Preserved default behavior: if specificTableList is empty, all tables/views are included.

Josel099 committed:
Refactor : refactor getAllTablesAndViews to filter tables at the SQL level

- Modify query to use a WHERE clause with an IN operator for filtering specific tables.
- Remove in-code filtering logic; rely on the database to return only the desired results.
-update readme.
